### PR TITLE
refactor: `deriveAddressSeed` and `deriveAddress` utilities

### PR DESCRIPTION
**Pull request by vadorovsky**
**Problems**

- The distinction between `derive_address_seed` and `derive_address` was
  unclear and we were inconsistent in it:
  - We ended up applying address Merkle tree public key in both
    functions, which is confusing.
- Before this change, there was no TypeScript function for deriving
  **address seed**. There was only `deriveAddress`, but deriving the
  unified seed was a mystery for developers.
- We have two utilities for hashing and truncating to BN254:
  - `hash_to_bn254_field_size_be` - the older one, which:
    - Searches for a bump in a loop, adds it to the hash inputs and then
      truncates the hash. That doesn't make sense, because truncating
      the hash should be sufficient, adding a bump is unnecessary.
    - Another limitation is that it takes only one sequence of bytes,
      making it difficult to provide multiple inputs without
      concatenating them.
  - `hashv_to_bn254_field_size` - the newer one, which:
    - Just truncates the hash result, without the bump mechanism.
    - Takes 2D byte slice as input, making it possible to pass multiple
      inputs.

**Changes**

- Don't add MT pubkey in `derive_address_seed`. It's not a correct place
  for it to be applied. The distinction between `derive_address_seed`
  and `derive_address` should be:
  - `derive_address_seed` takes provided seeds (defined by the
    developer) and hashes them together with the program ID. This
    operation is done only in the third-party program.
  - `derive_address` takes the address seed (result of
    `address_address_seed`) and hashes it together with the address
    Merkle tree public key. This is done both in the third-party program
    and in light-system-program. light-system-program does that as a
    check whether the correct Merkle tree is used.
- Adjust the stateless.js API:
  - Provide `deriveAddressSeed` function.
  - Add unit tests, make sure that `deriveAddressSeed` and
    `deriveAddress` provide the same results as the equivalent functions
    in Rust SDK.

### CODE REVIEW

**Review by SwenSchaeferjohann — COMMENTED** [review]
lgtm, just needs the name_service program adjusted

File: sdk/src/address.rs [hunk 1]
```diff
@@ -66,19 +66,12 @@ pub fn pack_new_address_params(
 ///     &address_merkle_context,
```
Comment (ananas-block) [inline sdk/src/address.rs:66]: Comment is not adapted

Comment (vadorovsky) [reply sdk/src/address.rs]: Good catch, done

File: js/stateless.js/src/utils/address.ts [hunk 2]
```diff
@@ -12,13 +21,13 @@ import { getIndexOrAdd } from '../instruction';
  *                          defaultTestStateTreeAccounts().merkleTree
  * @returns                 Derived address
  */
-export async function deriveAddress(
+export function deriveAddress(
     seed: Uint8Array,
     merkleTreePubkey: PublicKey = defaultTestStateTreeAccounts().merkleTree,
```
[the remark this answers is not in the document]
Comment (vadorovsky) [reply js/stateless.js/src/utils/address.ts:26]: Yes, it should, good catch.

[the remark this answers is not in the document]
Comment (me-imfhd) [reply js/stateless.js/src/utils/address.ts]: Nice, btw the comment is not consistent in the next commit

[the remark this answers is not in the document]
Comment (vadorovsky) [reply js/stateless.js/src/utils/address.ts]: Right. I also updated the variable name now to make it crystal clear it's address Merkle tree.

**Review by SwenSchaeferjohann — APPROVED** [review]

**Pull request merged**